pkg/controller/webhookbootstrap/controller.go

```diff
@@ -0,0 +1,488 @@
+	if err != nil {
+		return nil, err
+	}
+	return crtData, nil
+}
+
+// Register registers and constructs the controller using the provided context.
+// It returns the workqueue to be used to enqueue items, a list of
+// InformerSynced functions that must be synced, or an error.
+func (c *controller) Register(ctx *controllerpkg.Context) (workqueue.RateLimitingInterface, []cache.InformerSynced, error) {
+	// construct a new named logger to be reused throughout the controller
+	//log := logf.FromContext(ctx.RootContext, ExperimentalControllerName)
+
+	// create a queue used to queue up items to be processed
+	queue := workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(time.Second*5, time.Minute*30), ControllerName)
+
+	// obtain references to all the informers used by this controller
+	secretsInformer := ctx.KubeSharedInformerFactory.Core().V1().Secrets()
+
+	// build a list of InformerSynced functions that will be returned by the Register method.
+	// the controller will only begin processing items once all of these informers have synced.
+	mustSync := []cache.InformerSynced{
+		secretsInformer.Informer().HasSynced,
+	}
+
+	// set all the references to the listers for used by the Sync function
+	c.secretLister = secretsInformer.Lister()
+
+	// register handler functions
+	secretsInformer.Informer().AddEventHandler(&controllerpkg.QueuingEventHandler{Queue: queue})
+
+	c.kubeClient = ctx.Client
+
+	c.webhookDNSNames = ctx.WebhookBootstrapOptions.DNSNames
+	c.webhookCASecret = ctx.WebhookBootstrapOptions.CASecretName
+	c.webhookServingSecret = ctx.WebhookBootstrapOptions.ServingSecretName
+	c.webhookNamespace = ctx.WebhookBootstrapOptions.Namespace
+	c.certificateNeedsRenew = ctx.IssuerOptions.CertificateNeedsRenew
+	c.generatePrivateKeyBytes = generatePrivateKeyBytesImpl
+	c.signCertificate = signCertificateImpl
+
+	return queue, mustSync, nil
+}
+
+func (c *controller) ProcessItem(ctx context.Context, key string) error {
+	ctx = logf.NewContext(ctx, nil, ControllerName)
+	log := logf.FromContext(ctx)
+
+	if len(c.webhookDNSNames) == 0 {
+		log.Info("No webhook DNS names provided on start-up, not processing any resources.")
+		return nil
+	}
+
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		log.Error(err, "error parsing resource key in queue")
+		return nil
+	}
+
+	if c.webhookNamespace != namespace || !(c.webhookCASecret == name || c.webhookServingSecret == name) {
+		return nil
+	}
+
+	secret, err := c.secretLister.Secrets(namespace).Get(name)
+	if apierrors.IsNotFound(err) {
+		log.Info("secret resource no longer exists", "key", key)
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	switch name {
+	case c.webhookCASecret:
+		return c.syncCASecret(ctx, secret)
+	case c.webhookServingSecret:
+		return c.syncServingSecret(ctx, secret)
+	}
+
+	// TODO: set up scheduledWorkQueue for auto-renewal
```

Comment:
As we spoke about on the call we should add a scheduled re-queue in a time based on the standard renew before